### PR TITLE
module2 rewiev

### PR DESCRIPTION
**Pull request by maksimrut**

### CODE REVIEW

File: model/src/main/java/com/epam/esm/repository/BaseRepository.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * The interface Base repository.
+ *
+ * @param <T> the type parameter
+ * @param <E> the type parameter
+ *
+ * @author Maksim Rutkouski
+ */
+public interface BaseRepository<T, E> {
+
```
Comment (gshkvr) [inline model/src/main/java/com/epam/esm/repository/BaseRepository.java:15]: Seems that first generic value is always Long, so you can get rid of it

**Comment by gshkvr** [comment]
All changes look good to me. One note is that you don't validate incoming parameters

File: model/src/main/java/com/epam/esm/repository/BaseRepository.java (hunk 1 above)
Comment (maksimrut) [reply model/src/main/java/com/epam/esm/repository/BaseRepository.java:15]: I agree. In case of my project it's really not necessary. Should I correct it?

**Comment by maksimrut** [comment]
Thanks for your rewiev and note. I thought it's woudle be enough to validate data in entity classes with annotations (@NotBlank, @Size, etc.) for objects creating. If I have incorrect data at endpoints, I just handle an exception. We don't have frontend in module 2, I don't use dto in project (model mapper instead), so I decided implement only validation with annotations. Another validation on imaginary frontend. If I should make some corrections, please write to me. Thanks

File: model/src/main/java/com/epam/esm/entity/GiftCertificate.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+    @NotNull(message = "Certificate duration should not be null")
+    @Min(value = 1, message = "Certificate duration should not be less than 1")
+    @Max(value = 365, message = "Certificate duration should be less than 365 days")
+    private Integer duration;
+
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime createDate;
+
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
```
Comment (ZhukovAnton) [inline model/src/main/java/com/epam/esm/entity/GiftCertificate.java:35]: Better to get such constants from pre-configured Constants classes or from the properties

File: model/src/main/java/com/epam/esm/entity/GiftCertificate.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+import com.fasterxml.jackson.annotation.JsonFormat;
+import jakarta.validation.Valid;
+import jakarta.validation.constraints.*;
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.HashSet;
+import java.util.Set;
+
+public class GiftCertificate extends BaseEntity {
```
Comment (ZhukovAnton) [inline model/src/main/java/com/epam/esm/entity/GiftCertificate.java:12]: Lombok annotation `@Data` will be a good choice for entities, when you need both getters and setters + equals and hashcode implementations

**Review by ZhukovAnton — APPROVED** [review]

File: model/src/main/java/com/epam/esm/entity/GiftCertificate.java (hunk 3 above)
Comment (maksimrut) [reply model/src/main/java/com/epam/esm/entity/GiftCertificate.java:12]: Thank you for valuable notes!

**Pull request merged**